presto-hive/src/main/java/com/facebook/presto/hive/PrestoS3Client.java

```diff
@@ -0,0 +1,178 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive;
+
+import com.amazonaws.ClientConfiguration;
+import com.amazonaws.Protocol;
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.auth.AWSStaticCredentialsProvider;
+import com.amazonaws.auth.BasicAWSCredentials;
+import com.amazonaws.auth.InstanceProfileCredentialsProvider;
+import com.amazonaws.metrics.RequestMetricCollector;
+import com.amazonaws.regions.Region;
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3Builder;
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.facebook.presto.hive.s3.HiveS3Config;
+import com.facebook.presto.hive.s3.PrestoS3FileSystemMetricCollector;
+import com.facebook.presto.hive.s3.PrestoS3FileSystemStats;
+import io.airlift.units.Duration;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+
+import java.net.URI;
+import java.util.Optional;
+
+import static com.amazonaws.client.builder.AwsClientBuilder.EndpointConfiguration;
+import static com.amazonaws.regions.Regions.US_EAST_1;
+import static com.facebook.presto.hive.s3.S3ConfigurationUpdater.S3_ENDPOINT;
+import static com.facebook.presto.hive.s3.S3ConfigurationUpdater.S3_PIN_CLIENT_TO_CURRENT_REGION;
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static com.google.common.base.Verify.verify;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+
+class PrestoS3Client
+{
+    private static final String S3_ACCESS_KEY = "presto.s3.access-key";
+    private static final String S3_SECRET_KEY = "presto.s3.secret-key";
+    private static final String S3_CREDENTIALS_PROVIDER = "presto.s3.credentials-provider";
+    private static final String S3_USE_INSTANCE_CREDENTIALS = "presto.s3.use-instance-credentials";
+    private static final String S3_CONNECT_TIMEOUT = "presto.s3.connect-timeout";
+    private static final String S3_SOCKET_TIMEOUT = "presto.s3.socket-timeout";
+    private static final String S3_SSL_ENABLED = "presto.s3.ssl.enabled";
+    private static final String S3_MAX_ERROR_RETRIES = "presto.s3.max-error-retries";
+    private static final String S3_USER_AGENT_PREFIX = "presto.s3.user-agent-prefix";
+    private static final String S3_SELECT_PUSHDOWN_MAX_CONNECTIONS = "hive.s3select-pushdown.max-connections";
+    private static String s3UserAgentSuffix = "presto";
+    static AmazonS3 s3Client;
+
+    PrestoS3Client(Configuration config, HiveClientConfig clientConfig)
+    {
+        if (s3Client == null) {
+            synchronized (PrestoS3Client.class) {
```

Comment:
Another suggestion perhaps worth exploring.  Could we have a singleton that is registered in a `Module`, which is a supplier/provider of `PrestoS3Client`, and we thread this down to `S3SelectRecordCursorProvider`, which then retrieve the `PrestoS3Client` directly?  We would avoid the package-private static variable that way.  It would also reduce the plumbing for `HiveClientConfig` (that would just go into the supplier/provider).  It would also simplify this class and its subclass--it seems like this superclass is used for memoization and configuration, whereas the subclass is used for custom logic.  It seems like the configuration/memoization concerns can be factored out into a supplier/provider and completely separated out from `PrestoS3SelectClient`'s inheritance hierarchy.

Reply:
@sameer2086 for double check pattern to work, you would need to make `static s3Client` also `volatile`.
And even this would be a problem if someone has 2 hive connectors eg talking to different s3 regions or with different credentials. See @electrum 's https://groups.google.com/d/msg/presto-users/84As2sVuanc/JEXExk7fBQAJ

Reply:
Going Guice way may solve both problems.

Reply:
Yup agreed. The new revision contains that change.